src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -1083,11 +1081,12 @@ internal void LogBuffer(int size)
 
             if (_bufferList != null)
             {
-                for (int i = 0; i < _bufferListInternal!.Count; i++)
+                ReadOnlySpan<ArraySegment<byte>> buffers = CollectionsMarshal.AsSpan(_bufferListInternal);
+                for (int i = 0; i < buffers.Length; i++)
                 {
-                    WSABuffer wsaBuffer = _wsaBufferArrayPinned![i];
-                    NetEventSource.DumpBuffer(this, wsaBuffer.Pointer, Math.Min(wsaBuffer.Length, size));
-                    if ((size -= wsaBuffer.Length) <= 0)
+                    ref readonly ArraySegment<byte> segment = ref buffers[i];
+                    NetEventSource.DumpBuffer(this, Marshal.UnsafeAddrOfPinnedArrayElement(segment.Array!, segment.Offset), Math.Min(segment.Count, size));
```

Comment:
We do not need to use unsafe code here.  I'm also not sure it's valid... is the array guaranteed to be here?  Seems like this code is reachable from the async completion path.

Reply:
Assuming its not used in Parallel; then yes as it pins when `_bufferListInternal` is set in `BufferList { set; }`